### ISSUE TITLE
nc_test fails when optimization greater than -O0 is invoked.

### ISSUE THREAD

**Issue by WardF** [Unidata/netcdf-c#1983]
A failure in `nc_test/nc_test` was initially reported on the `ppc64le` platform as part of the tests for `conda-forge/libnetcdf-feedstock` release of 4.8.0.  Subsequent testing has revealed an issue with some, but not all, compilers.  More details to follow, opening an issue to capture this.

## Update 1

Issue has been observed on M1 hardware, both in OSX, and emulated ppc64le (via `qemu`).  Issue does *not* manifest on M1 hardware in virtualized ubuntu, or on intel-based OSX.  I'll update with compiler versions (which is a big piece of the puzzle).

**Comment by dopplershift** [comment]
I can confirm that replacing `-O3` with `-O0` over in conda-forge/libnetcdf-feedstock#119 allowed the failing test to pass.

**Comment by dopplershift** [comment]
Some relevant build env information from over there. Host env:
```
    _libgcc_mutex:   0.1-conda_forge                conda-forge
    _openmp_mutex:   4.5-1_gnu                      conda-forge
    bzip2:           1.0.8-h4e0d66e_4               conda-forge
    c-ares:          1.17.1-h4e0d66e_1              conda-forge
    ca-certificates: 2020.12.5-h1084571_0           conda-forge
    curl:            7.76.0-h86bdb50_0              conda-forge
    hdf4:            4.2.13-h9a582f1_3              defaults   
    hdf5:            1.10.6-mpi_mpich_h588be55_1014 conda-forge
    jpeg:            9d-h339bb43_0                  conda-forge
    krb5:            1.17.2-h2004cfc_0              conda-forge
    libcurl:         7.76.0-hf726fd8_0              conda-forge
    libedit:         3.1.20191231-h41a240f_2        conda-forge
    libev:           4.33-h6eb9509_1                conda-forge
    libgcc-ng:       9.3.0-hc7a5eb4_18              conda-forge
    libgfortran-ng:  9.3.0-hb3da1e1_18              conda-forge
    libgfortran5:    9.3.0-hb3da1e1_18              conda-forge
    libgomp:         9.3.0-hc7a5eb4_18              conda-forge
    libnghttp2:      1.43.0-h42039ad_0              conda-forge
    libssh2:         1.9.0-ha5a9321_6               conda-forge
    libstdcxx-ng:    9.3.0-h22bc1b8_18              conda-forge
    mpi:             1.0-mpich                      conda-forge
    mpich:           3.4.1-hb1f5ef7_104             conda-forge
    ncurses:         6.2-hea85c5d_4                 conda-forge
    openssl:         1.1.1k-h4e0d66e_0              conda-forge
    tk:              8.6.10-h38e1d09_1              conda-forge
    zlib:            1.2.11-h6eb9509_1010           conda-forge
```

Build env:
```
    _libgcc_mutex:                 0.1-conda_forge         conda-forge
    _openmp_mutex:                 4.5-1_gnu               conda-forge
    binutils_impl_linux-ppc64le:   2.35.1-h5836da8_2       conda-forge
    binutils_linux-ppc64le:        2.35-hcfcb53e_30        conda-forge
    bzip2:                         1.0.8-h4e0d66e_4        conda-forge
    c-ares:                        1.17.1-h4e0d66e_1       conda-forge
    ca-certificates:               2020.12.5-h1084571_0    conda-forge
    cmake:                         3.20.1-hf140c13_0       conda-forge
    expat:                         2.3.0-h3b9df90_0        conda-forge
    gcc_impl_linux-ppc64le:        9.3.0-ha8fcf76_18       conda-forge
    gcc_linux-ppc64le:             9.3.0-h2662597_30       conda-forge
    gxx_impl_linux-ppc64le:        9.3.0-h6db6fac_18       conda-forge
    gxx_linux-ppc64le:             9.3.0-hdc9eec7_30       conda-forge
    kernel-headers_linux-ppc64le:  3.10.0-hbe0c576_10      conda-forge
    krb5:                          1.17.2-h2004cfc_0       conda-forge
    ld_impl_linux-ppc64le:         2.35.1-ha35d02b_2       conda-forge
    libcurl:                       7.76.0-hf726fd8_0       conda-forge
    libedit:                       3.1.20191231-h41a240f_2 conda-forge
    libev:                         4.33-h6eb9509_1         conda-forge
    libgcc-devel_linux-ppc64le:    9.3.0-h9f60fdd_18       conda-forge
    libgcc-ng:                     9.3.0-hc7a5eb4_18       conda-forge
    libgomp:                       9.3.0-hc7a5eb4_18       conda-forge
    libnghttp2:                    1.43.0-h42039ad_0       conda-forge
    libssh2:                       1.9.0-ha5a9321_6        conda-forge
    libstdcxx-devel_linux-ppc64le: 9.3.0-h7901c9a_18       conda-forge
    libstdcxx-ng:                  9.3.0-h22bc1b8_18       conda-forge
    libuv:                         1.41.0-h4e0d66e_0       conda-forge
    lz4-c:                         1.9.3-h3b9df90_0        conda-forge
    make:                          4.3-hf817498_1          conda-forge
    ncurses:                       6.2-hea85c5d_4          conda-forge
    openssl:                       1.1.1k-h4e0d66e_0       conda-forge
    pkg-config:                    0.29.2-h339bb43_1008    conda-forge
    rhash:                         1.4.1-h4e0d66e_0        conda-forge
    sysroot_linux-ppc64le:         2.17-h8b29623_10        conda-forge
    tk:                            8.6.10-h38e1d09_1       conda-forge
    xz:                            5.2.5-h6eb9509_1        conda-forge
    zlib:                          1.2.11-h6eb9509_1010    conda-forge
    zstd:                          1.4.9-h65c4b1a_0        conda-forge
```

**Comment by WardF** [comment]
Surprisingly, no failure on a raspberry pi ARM device (gcc 10.2).

**Comment by edwardhartnett** [comment]
I would suggest you try on standard unix box with valgrind, and with address sanitizer.

**Comment by WardF** [comment]
The address sanitizer hadn't turned anything up, but valgrind was a good idea.  Valgrind detected a few undefined variables that could have been the issue, but addressing them in the source `.m4` files hasn't made a difference to the observed error (although valgrind no longer reports any errors).  Stepping through in lldb now.

**Comment by c-ney** [comment]
This (or a similar) issue has recently popped up in Opensuse Tumbleweed i586, x86_64 and aarch64 (some logs at [0]) when they switched gcc from 10 to 11 by default (it was a huge update, so might be coming from something else, but this seems the most noteworthy change in my eyes). I only tested x86_64 locally, compiling with `-O0` or `-O1` works fine, but `-O2` will result in nc_test failing.

[test-suite.log](https://github.com/Unidata/netcdf-c/files/6626537/test-suite.log)

Haven't had the time to look any further yet.

[0] https://build.opensuse.org/package/show/science/netcdf

**Comment by WardF** [comment]
Thanks for percolating this back up; I'll pick this back up and see if I can figure out what's going on.

**Comment by c-ney** [comment]
Egbert Eich from the SUSE team was able to pin this down to strict aliasing requirements in GCC11. After disabling it in [0], the package (now 4.8.0) builds with successful tests [1].

[0] https://build.opensuse.org/request/show/904468
[1] https://build.opensuse.org/package/show/science/netcdf

**Comment by WardF** [comment]
Wow, thank you, I've been working on this and a different issue in an effort to get our next release out, this is a great help. I will sort through his results and get this addressed!

**Comment by DennisHeimbigner** [comment]
What is the exact fix: -Wnomaybe-unitialized ?

**Comment by WardF** [comment]
@DennisHeimbigner I believe the temporary fix, pending the broader evaluation of our aliasing, is to pass `-fno-strict-aliasing` for gcc >= 11.  Testing this fix now.

**Comment by WardF** [comment]
@DennisHeimbigner still encountering the issue on OSX w/ gcc, so continue to move forward working on this.

**Comment by edwardhartnett** [comment]
Can we fix the code instead of adding a flag? What is the exact code issue that is causing the problem, do we know?

**Comment by WardF** [comment]
@edhartnett Fixing the code would be ideal, but the code, and the fix, still need tbd.  We need to get 4.8.1 out the door and if a flag will let us implement a work around until the fix is determined, that may be the route we end up taking. But we're still sorting things out on our end.

**Comment by ArchangeGabriel** [comment]
On ArchLinux (x86_64), we have the same issue but I can confirm that `-fno-strict-aliasing` solves it for us.

**Comment by WardF** [comment]
This error has been temporarily addressed w/ the configuration flags so that it is no longer blocking `4.8.1`, and a 'proper' fix will be investigated for `4.8.2`.

**Comment by WardF** [comment]
Thanks to the renewed interest, some additional info/reading provided by @dopplershift, and an expanded knowledge of `-fsanitize`, I think we have a fix for this.  It also looks like we have a lot of additional work to do cleaning up some operations that are undefined, but consistent.  But first things first.  I'll be testing the potential fix for this shortly.
